src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/jvm/ThreadList.java

```diff
@@ -104,44 +111,75 @@ public String toString() {
+   */
   public static Map<Long, ThreadState> getNativeTidMap() {
-    synchronized (ThreadList.class) {
-      if (System.currentTimeMillis() > lastRunTime + minRunInterval) {
-        runThreadDump(pid, new String[0]);
+    if (vmAttachLock.tryLock()) {
+      try {
+        // Thread dumps are expensive and therefore we make sure that at least
+        // minRunInterval milliseconds have elapsed between two attempts.
+        if (System.currentTimeMillis() > lastRunTime + minRunInterval) {
+          runThreadDump(pid, new String[0]);
+        }
+      } finally {
+        vmAttachLock.unlock();
       }
-      // - sending a copy so that if runThreadDump next iteration clears it; caller still has the
-      // state at the call time
-      // - not too expensive as this is only being called from Scheduled Collectors (only once in
-      // few seconds)
-      return new HashMap<>(nativeTidMap);
+    } else {
+      StatsCollector.instance().logException(StatExceptionCode.JVM_ATTACH_LOCK_ACQUISITION_FAILED);
     }
+
+    // - sending a copy so that if runThreadDump next iteration clears it; caller still has the
+    // state at the call time
+    // - not too expensive as this is only being called from Scheduled Collectors (only once in
+    // few seconds)
+    return new HashMap<>(nativeTidMap);
   }
 
+  /**
+   * This method is called from the critical bulk and search paths which PA
+   * intercepts. This method used to try to do a thread dump if it could not
+   * find the information about the thread in question. The thread dump is an
+   * expensive operation and can stall see VirtualMachineImpl#VirtualMachineImpl()
+   * for jdk-11 u06. We don't want the ES threads to pay the price. We skip this
+   * iteration and then hopefully in the next call to getNativeTidMap(), the
+   * OSMetricsCollector#collectMetrics will fill the jTidMap. This transfers the
+   * responsibility from the ES threads to the PA collector threads.
+   *
+   * @param threadId The threadId of the current thread.
+   * @return If we have successfully captured the ThreadState, then we emit it or Null otherwise.
+   */
   public static ThreadState getThreadState(long threadId) {
     ThreadState retVal = jTidMap.get(threadId);
-
-    if (retVal != null) {
-      return retVal;
-    }
-
-    synchronized (ThreadList.class) {
-      retVal = jTidMap.get(threadId);
-
-      if (retVal != null) {
-        return retVal;
-      }
-
-      runThreadDump(pid, new String[0]);
+    if (retVal == null) {
+      StatsCollector.instance().logException(StatExceptionCode.NO_THREAD_STATE_INFO);
     }
-
-    return jTidMap.get(threadId);
+    return retVal;
   }
 
   // Attach to pid and perform a thread dump
   private static void runAttachDump(String pid, String[] args) {
     VirtualMachine vm = null;
     try {
       vm = VirtualMachine.attach(pid);
+    } catch (AttachNotSupportedException ansEx) {
+      if (ansEx.getMessage().contains("java_pid")) {
+        LOGGER.debug(
```

Comment:
Shouldn't this be Logger.error?

Reply:
You are right but if attach fails consecutively, it might generate a lot of logs. Therefore, added as debug so that it emits in test env. It also emits a metric that should be tracked for in prod failures of this nature.